### ISSUE TITLE
Improving the remote runner framework

### ISSUE THREAD

**Issue by famz** [avocado-framework/avocado#1612]
The "avocado on remote" problem
===

Currently, trying to run tests in a "stock" docker image won't work:

```
$ avocado run pass --docker-cmd "sudo docker" --docker fedora:latest
DOCKER     : Container id '68f6ab0c7defd50f64f6fdaa2947907c561cebc446c6e33c39250e04c900f37a'

Avocado job failed: JobError: Remote machine does not seem to have avocado installed
```

This is apparently because the specified fedora image doesn't have avocado installed - the current remote runner API requires that, which is a reasonable design decision as that makes it easy to support running **any test** remotely. In other words, tests can be written **orthogonally** to the type of the remote it will later run, which is admittedly a valuable decoupling strategy.

However, this requirement is not always practical to fullfill, in a more general sense -- the remote may not have Python, or is so specialized that installing avocado is difficult, or the intended tests are so apparently independent of avocado itself (that is not so uncommon, IMO), that user would just copy and run the test script manually, rather than bothering installing avocado and use the remote plugin.

How avocado can help, in such cases? My proposal here is to enhance the remote runner framework  to allow test cases to be written in a such a way that avocado at remote side is not strictly necessary.

_This is also a nice advantage of Ansible over its many competitors - it only requires ssh._

That can be made possible in two ways:

Approach A

---

A test is written **independently from avocado**. This way it can be simply copied over to the remote with `send_files()` method provided by the remote plugin, and in turn executed with the `run()` method. The local avocado process manages the job; all results and output are collected remotely.

Approach B

---

A test is written with the knowledge of remote(s) in mind. When job begins, the test can get remote information of the job (e.g. get the created remoter object) from the avocado core, and perform test steps remotely, by calling `run()` methods on the remoter object, or another supported remote method, etc..

A test that inherently involves multiple machines can even explicitly accept or require multiple remotes. For instances, such a test can be a QEMU migration test between two different hosts, or a network test that wants to test two hosts communicating with each other, or a large scale test that can run on a cluster of an arbitrary number of nodes (more on this later).

How to coexist with the current "any test can run on any remote" idea?

---

By default the job still checks that the remote has avocado installed to run usual tests remotely, unless the job only has tests of either type above. Consequently, we must have a way to annotate or label tests so the job runner can identify them. Approach B is easy because it is implied the test code is using avocado framework; for appraoch A, I think adding a recognizable magic comment would be good enough.

More generic remote option syntax
===

Taking above proposal one more step further, we can make some changes to the command line interface with regarding to how remotes are specified, to allow more interesting use cases.

Multiple instances of remotes

---

In above mentioned example multi-remote use cases, a test may take two remotes as the input and operate them to perform test steps, with the same type or not. The current per-plugin ad-hoc command line interface (e.g. `--remote-hostname`) is not expressive enough to achive that. Analogue to this in the QEMU project is the legacy "-hda" option which is replaced by the more generic "-device ide-hd,..." syntax.

From a user and test writer PoV, I imaging specifying a remote to run my tests like this:

```
avocado run pass --remote docker,docker-cmd="sudo docker",image=fedora
```

and specifying multiple remotes as necessary like this:

```
# post-copy-migrate test will create VMs on "source" machine and migrate it to "target".
# In this case, a guest created on HOST1 wille be migrated as an L2 guest created inside a local L1 guest, called VM_DOMAIN
avocado run post-copy-migrate \
    --remote ssh,id=source,hostname=HOST1 \
    --remote vm,id=target,domain=VM_DOMAIN
```

Stacking remote plugins

---

_Warning: this is a crazy idea, hold tight._

We support remotely running over SSH, as well as remotely running in docker, can we support running in a docker instance on a remote machine, or docker in a VM, or VM in a VM? I think we can achive this flexibility if we write all remote plugins in a careful way that they don't assume they are running locally, instead they should operates on another remote. For example, the docker plugin can call the underlying remote object's `run()` method from its own `run()` method. Of course, the recursion has to terminate somewhere, which would be a special "local" remote plugin that runs command with `subprocess`, or a `fabric` remote plugin that calls `fabric` functions; all other plugins should try to be "stackable" on each other.

The following example demostrates another L1 to L2 migration test happening across two hosts:

```
avocado run post-copy-migrate \
    --remote ssh,id=source,hostname=HOST1 \
    --remote ssh,id=target-host,hostname=HOST2 \
    --remote vm,base=target-hostid=target,domain=VM_DOMAIN
```

Note: how `post-copy-migrate` will establish communication between `source` and `target` remains unanswered, and apparently needs further discussion. As the simplest solution, a `get_public_ip()` method can be implemented by all remote plugins, so the test can try to connect from one to another. Taking the above example, what is returned by `get_public_ip()` for the `target` remote must be reachable from `source`, in order for the test to succeed.

Comments?

**Comment by ldoktor** [comment]
Hello @famz

lots of interesting ideas here. I think it'd be nice if you re-send this as RFC in the mailing-list to simplify splitting it in parts... Anyway let me comment some ideas here.


The Approach A
--------------

currently it's not really possible as the remote runner executes the whole job remotely. And job is an avocado concept, therefor it requires Avocado installed.

Anyway I tackled some of the areas you are mentioning in the multi-stream RFC. I think it'd be nice to change the remote execution to run per-test, rather than the job, therefor the job would live on the main machine and the remote machine would only start the test and report results. Then it'd be possible to modify SimpleTests in order to execute them remotely without the need of python. It'd not be just copy and run, because it is still test, it requires initialization and some postprocess and those steps are Avocado&python.

Obviously it would still be impossible to run Avocado tests without Avocado, because Avocado is not just the runner and CLI. It's also the utils and libraries and they are used inside the tests, but we don't need the full-sized Avocado. We could use lightweight standalone test execution with only avocado utils and basic workflow supported. We'd lose the possibility to use gdb/wrappers and other nice features of Avocado, but we don't support them remotely even now.


The Approach B
--------------

Strictly speaking this is no matter for Avocado. Eventually we might document ways to do it, even provide `avocado.utils` to help writing tests, but what's inside test is inside test. You can create a machine, use Remoter/fabric/aexpect and do whatever you need. For example there is nothing what would prevent you from opening two `aexpect.ShellSession("ssh ...")` sessions and run `avocado run start_migrate --vt-extra-params ip=test2` and `avocado run receive_migrate` and control the flow from the main machine. Anyway we both see it's not really optimal.


Approach C
----------

Let's use our buzz word, Job API. It could be possible to create a SimpleRemote test, which would use eg. aexpect.ShellSession to connect to a remote machine and allow writing simple tests driven from the local host. The problem is what happens when we lose connection... That's why I always prefer a runner (even if simplified) on the remote machine which can drive the execution including the `tearDown` part where the network can be restored in order to call back and report the results.


Synchronization
---------------

Basically you one can use python standard libraries for synchronization, or autotest barriers, or we can come up with another solution. ~ half a year ago I sent a demonstration of barriers mechanism which used the multiplexer to specify the ip:port parameters, but that's just an example.


Multiplex instances of remote
-----------------------------

Currently not possible as that would require creating multiple jobs. We're not running tests, but we are running jobs remotely. Again with Job API it should be eventually possible to create a script with a loop where you create several jobs and run them in parallel and each job could use different remote plugin.

As for the `--remote $type,$args` as an universal API I like it. It could be implemented even now by creating `remote` plugin type, which would register types of the remotes and avocado would forward the params to the corresponding plugin. The limitation still applies, unless we allow the basic runner to run multiple jobs we'd only have to support one remote at a time.

Anyway I suggest sending this as an RFC so we can easily reply and perhaps some ideas can be extracted, or used by Cleber during the multi-test RFC he's resurresting in this sprint.
Lukáš

**Comment by clebergnu** [comment]
@famz since the multi stream proposal is related (not identical) to your ideas or use cases, I thought I'd post a link here: https://www.redhat.com/archives/avocado-devel/2017-April/msg00002.html

**Comment by ldoktor** [comment]
@famz this proposal is with us for quite a while but we don't really have any plan on working on this any time soon (we want to tackle the multi-stream, though). Do you plan on preparing something or can we close this as wontfix (until someone finds the time?).

**Comment by famz** [comment]
I don't have time for this in the near future as well. So closing.
